File: sdk/python/examples/06_motor_control.py
"""电机操作全流程：诊断、清错、标零、增益监控。"""

import time
from brainstem_sdk import ThunderClient

JOINT_NAMES = [
    "FR Hip", "FR Thigh", "FR Calf", "FR Foot",
    "FL Hip", "FL Thigh", "FL Calf", "FL Foot",
    "RR Hip", "RR Thigh", "RR Calf", "RR Foot",
    "RL Hip", "RL Thigh", "RL Calf", "RL Foot",
]

dog = ThunderClient("192.168.66.190")
try:
    # ──── 1. 电机诊断 ────────────────────────────────────

    print("=== 电机诊断 ===")
    motors = dog.get_motor_status()

    online_count = sum(1 for m in motors if m.online)
    fault_count = sum(1 for m in motors if m.errors)
    print(f"在线: {online_count}/16, 故障: {fault_count}/16")

    for m in motors:
        status = "OK" if m.online and not m.errors else "FAULT"
        print(
            f"  [{m.id:2d}] {JOINT_NAMES[m.id]:10s}  "
            f"{status:5s}  "
            f"temp={m.temperature:4.0f}C  "
            f"V={m.voltage:5.1f}  "
            f"pos={m.position:+7.3f}rad  "
            f"vel={m.velocity:+7.3f}  "
            f"torque={m.torque:+6.2f}Nm"
        )

    # ──── 2. 电压检查 ────────────────────────────────────

    print("\n=== 电压检查 ===")
    voltages = dog.get_voltage()
    min_v = min(v for v in voltages if v > 0) if any(v > 0 for v in voltages) else 0
    max_v = max(voltages)
    print(f"电压范围: {min_v:.1f}V ~ {max_v:.1f}V")
    if min_v < 42.0:
        print(f"  WARNING: 最低电压 {min_v:.1f}V < 42.0V，请充电!")
    else:
        print(f"  电压正常 (阈值 42.0V)")

    # ──── 3. 故障处理 ────────────────────────────────────

    if fault_count > 0:
        print(f"\n=== 故障处理 ===")
        faulted = [m.id for m in motors if m.errors]
        print(f"故障关节: {[JOINT_NAMES[i] for i in faulted]}")

        response = input("是否清除故障？(y/n): ")
        if response.lower() == "y":
            dog.clear_motor_fault(faulted)
            print("故障已清除。")
            time.sleep(0.5)

            # 重新检查
            motors2 = dog.get_motor_status()
            still_faulted = [m.id for m in motors2 if m.errors]
            if still_faulted:
                print(f"  仍有故障: {[JOINT_NAMES[i] for i in still_faulted]}")
                print("  请检查硬件连接或重新上电。")
            else:
                print("  所有故障已清除!")

    # ──── 4. 标零（可选）───────────────────────────────────

    print("\n=== 标零 ===")
    state = dog.get_state()
    if state == "grounded":
        response = input("当前在 Grounded 状态，是否标零？(y/n): ")
        if response.lower() == "y":
            dog.set_zero()
            print("标零完成。当前位置已设为零点。")
    else:
        print(f"当前状态: {state}，标零需要在 Grounded 状态。")

    # ──── 5. 实时关节监控 ────────────────────────────────

    print("\n=== 实时关节监控 (5 秒) ===")
    dog.enable()
    dog.stand_up()
    time.sleep(1)

    count = 0
    try:
        for joint in dog.listen_joint():
            name = JOINT_NAMES[joint.id] if joint.id < 16 else f"Joint{joint.id}"
            print(
                f"  {name:10s}  "
                f"pos={joint.position:+7.3f}  "
                f"vel={joint.velocity:+7.3f}  "
                f"torque={joint.torque:+6.2f}",
                end="\r",
            )
            count += 1
            if count > 250:  # ~5 seconds at 50Hz
                break
    except KeyboardInterrupt:
        pass

    print("\n\n监控结束。")
finally:
    try:
        dog.sit_down()
    except Exception:
        pass
    try:
        dog.disable()
    except Exception:
        pass
    dog.close()

File: sdk/cpp/examples/06_motor_control.cpp
// 06_motor_control.cpp — 电机操作全流程：诊断、清错、标零、实时关节监控
#include "orix/orix.h"

#include <algorithm>
#include <chrono>
#include <iomanip>
#include <iostream>
#include <thread>

static const char* JOINT_NAMES[16] = {
    "FR_hip",  "FR_thigh", "FR_calf",
    "FL_hip",  "FL_thigh", "FL_calf",
    "RR_hip",  "RR_thigh", "RR_calf",
    "RL_hip",  "RL_thigh", "RL_calf",
    "FR_foot", "FL_foot",  "RR_foot", "RL_foot",
};

int main() {
    orix::OrixClient dog("192.168.66.190");

    if (!dog.ping()) {
        std::cerr << "无法连接到机器人。\n";
        return 1;
    }

    try {
        // ── 1. 电机诊断 ───────────────────────────────────────
        std::cout << "=== 电机诊断 ===\n";
        auto motors = dog.get_motor_status();

        int online_count = 0, fault_count = 0;
        for (const auto& m : motors) {
            if (m.online) ++online_count;
            if (!m.errors.empty()) ++fault_count;
        }
        std::cout << "在线: " << online_count << "/16  "
                  << "故障: " << fault_count  << "/16\n\n";

        std::cout << std::fixed << std::setprecision(2);
        for (const auto& m : motors) {
            std::string st = (m.online && m.errors.empty()) ? "OK   " : "FAULT";
            std::cout
                << "  [" << std::setw(2) << m.id << "] "
                << std::setw(10) << std::left << JOINT_NAMES[m.id] << std::right
                << "  " << st
                << "  temp=" << std::setw(5) << m.temperature << "C"
                << "  V="    << std::setw(5) << m.voltage
                << "  pos="  << std::setw(7) << m.position  << "rad"
                << "  vel="  << std::setw(7) << m.velocity
                << "  trq="  << std::setw(6) << m.torque    << "Nm\n";
        }

        // ── 2. 电压检查 ───────────────────────────────────────
        std::cout << "\n=== 电压检查 ===\n";
        auto voltages = dog.get_voltage();
        double min_v = *std::min_element(voltages.begin(), voltages.end());
        double max_v = *std::max_element(voltages.begin(), voltages.end());
        std::cout << "电压范围: " << min_v << "V ~ " << max_v << "V\n";
        if (min_v < 42.0)
            std::cout << "  WARNING: 最低电压 " << min_v << "V < 42V，请充电!\n";
        else
            std::cout << "  电压正常 (阈值 42V)\n";

        // ── 3. 故障处理 ───────────────────────────────────────
        if (fault_count > 0) {
            std::cout << "\n=== 故障处理 ===\n";
            std::vector<int> faulted;
            for (const auto& m : motors)
                if (!m.errors.empty()) faulted.push_back(m.id);

            std::cout << "故障关节: ";
            for (int id : faulted) std::cout << JOINT_NAMES[id] << " ";
            std::cout << "\n清除故障中...\n";
            dog.clear_motor_fault(faulted);
            std::this_thread::sleep_for(std::chrono::milliseconds(500));

            auto motors2 = dog.get_motor_status();
            bool all_clear = true;
            for (const auto& m : motors2)
                if (!m.errors.empty()) { all_clear = false; break; }
            std::cout << (all_clear ? "  所有故障已清除!\n" : "  仍有故障，请检查硬件连接。\n");
        }

        // ── 4. 标零（须在 Grounded 状态）────────────────────
        std::cout << "\n=== 零点标定 ===\n";
        if (orix::to_string(dog.get_state()) == "grounded") {
            std::cout << "当前在 Grounded 状态，执行标零...\n";
            dog.set_zero();
            std::cout << "标零完成。\n";
        } else {
            std::cout << "当前状态非 Grounded，跳过标零。\n";
        }

        // ── 5. 实时关节监控（5 秒）───────────────────────────
        std::cout << "\n=== 实时关节监控（5 秒）===\n";
        dog.enable();
        dog.stand_up();
        std::this_thread::sleep_for(std::chrono::seconds(1));

        int count = 0;
        dog.listen_joint([&](const orix::JointData& j) -> bool {
            const char* name = j.id < 16 ? JOINT_NAMES[j.id] : "unknown";
            std::cout << "\r  [" << std::setw(2) << j.id << "] "
                      << std::setw(10) << std::left << name << std::right
                      << "  pos=" << std::setw(7) << j.position
                      << "  vel=" << std::setw(7) << j.velocity
                      << "  trq=" << std::setw(6) << j.torque
                      << std::flush;
            return ++count < 250; // ~5s at 50Hz
        });
        std::cout << "\n\n监控结束。\n";

    } catch (const orix::OrixError& e) {
        std::cerr << "错误: " << e.what() << "\n";
    }

    try { dog.sit_down(); } catch (...) {}
    try { dog.disable();  } catch (...) {}
    return 0;
}
